### PR TITLE
[torch.compile] Stop lazily compiling

### LINKED ISSUE #34034: [Bug]: vLLM-compile should not execute the decoder forward pass during compilation
### Your current environment

main

### 🐛 Describe the bug

During cold start compilation, vLLM-compile executes the text decoder forward pass using the max_num_batched_tokens size. I'm not completely sure how long a text decoder forward pass is, but if it is O(1s) then this is unnecessary and will save on compile time.

There's one complication in that I don't know when autotuning will happen. I hope that vLLM does warmup of a size before the first cudagraph capture (but also we do compile-time autotuning so I hope that works too...)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by zou3519**
Fixes https://github.com/vllm-project/vllm/issues/34034

Compile all ranges upfront in VllmBackend

Move Inductor compilation from lazy (first warmup execution) to eager (during `VllmBackend.__call__`). Previously, PiecewiseBackend compiled each range
on-demand when first encountered at warmup. Now, `VllmBackend.__call__` calls PiecewiseBackend.compile_all_ranges() for every subgraph before returning the
callable, so all compilation is complete before the first warmup run with real tensors.

This simplifies the compilation lifecycle:

- Remove check_for_ending_compilation and the _on_compilation_complete_callback context variable — cache saving and monitoring now happen directly at the end of `VllmBackend.__call__`
- Remove lazy compilation in _maybe_compile_for_range_entry — it now raises RuntimeError if a range wasn't compiled upfront (still supports the compiled_runnables path for AOT/cache loading)
- Simplify AOT save in decorators.py — call save_aot_compiled_function() directly after aot_compile() instead of going through a callback

### CODE REVIEW

File: tests/compile/test_compile_ranges.py [hunk 1]
```diff
@@ -73,6 +73,7 @@ def test_compile_ranges(use_fresh_inductor_cache):
             Range(start=16, end=16),
             Range(start=9, end=32),
             Range(start=64, end=64),
+            Range(start=128, end=128),
```
Comment (zou3519) [inline tests/compile/test_compile_ranges.py:76]: compile_sizes has 128 in it. Previously we would never trigger it because it was not one of batch sizes. Now we compile everything up front without seeing the batch sizes, so 128 is compiled as well.

File: vllm/compilation/piecewise_backend.py [hunk 2]
```diff
@@ -16,14 +15,59 @@
...
+                    )
+                    args.append(
+                        torch.zeros(
+                            new_shape,
+                            dtype=val.dtype,
+                            device=val.device,
+                        )
+                    )
+                else:
+                    args.append(val)
```
Comment (zou3519) [inline vllm/compilation/piecewise_backend.py]: TODO(rzou): this is wrong, need to create a new tensor here

File: vllm/compilation/piecewise_backend.py [hunk 3]
```diff
@@ -16,14 +15,47 @@
...
+    args: list[Any] = []
+    for node in graph.graph.nodes:
+        if node.op != "placeholder":
+            break
+        val = node.meta["example_value"]
+        if isinstance(val, torch.SymInt):
+            args.append(size)
+        elif isinstance(val, torch.Tensor):
+            new_shape = tuple(size if is_symbolic(d) else int(d) for d in val.shape)
+            args.append(torch.zeros(new_shape, dtype=val.dtype, device=val.device))
```
Comment (zhxchen17) [inline vllm/compilation/piecewise_backend.py]: Should empty() be faster for this case?
```suggestion
            args.append(torch.empty(new_shape, dtype=val.dtype, device=val.device))
```

File: vllm/compilation/piecewise_backend.py
Comment (zhxchen17) [inline vllm/compilation/piecewise_backend.py:379]: suggestion: after the code change, we will have 2 segments of time spent on "compile" and "profiling run" separately, instead of 1 combined time segment. So from user perspective this will create a hole in the logging where profiling run time doesn't get reported but vllm appears doing something in the background.

Sounds better to log the time passed for profiling run (after compile) as well.

**Review by zhxchen17 — COMMENTED** [review]
Thanks! This will fix the measurement for torch.compile time for both cold and warm start.

File: vllm/compilation/piecewise_backend.py (hunk 3 above)
Comment (zou3519) [reply vllm/compilation/piecewise_backend.py]: I wanted to avoid nondeterministic behavior with empty inputs

File: vllm/compilation/piecewise_backend.py
Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:379]: yup let me see how easy this is

Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:379]: I added it. the profiling run can include inductor autotuning time unfortunately

Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:379]: @zhxchen17 @ProExpertProg would you prefer the log look like..
```
torch.compile took Xs in total
Initial profiling run (may include torch.compile autotuning time) took Ys
```
or
```
torch.compile and initial profiling run took (X+Y)s in total
```
or something else?

The difference is we are measuring two stages:
- the first stage is just compilation (tracing, code generation, some minimal autotuning)
- the second stage is the warmup run where the actual forward pass is run. This is where torch.compile can do autotuning

It is not completely possible to move all of the autotuning from the warmup run to the compilation run at the moment

Comment (zhxchen17) [reply vllm/compilation/piecewise_backend.py:379]: I'm slightly inclined to the second option (aggregated report) until we find a way to move/ensure autotuning inside torch.compile(), since average probably doesn't know what auto tuning means.

Comment (ProExpertProg) [reply vllm/compilation/piecewise_backend.py:379]: I think given that the default is to not do any autotuning, can we just do:
```
torch.compile took Xs in total
Initial profiling run took Ys
# And if autotuning is enabled:
torch.compile autotuning took Zs
```

Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:379]: autotuning for torch.compile is actually always enabled. max-autotune just enables more autotuning

Comment (ProExpertProg) [reply vllm/compilation/piecewise_backend.py:379]: Can we measure that separately the profile run? Either way I prefer the separate logs

Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:379]: We are not able to measure the autotuning time separately at the moment (or at least I don't know how to do it lol)

Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:379]: I updated to two log messages

File: vllm/compilation/piecewise_backend.py [hunk 4]
```diff
@@ -16,14 +15,47 @@
...
+
+    args: list[Any] = []
+    for node in graph.graph.nodes:
+        if node.op != "placeholder":
+            break
+        val = node.meta["example_value"]
+        if isinstance(val, torch.SymInt):
+            args.append(size)
+        elif isinstance(val, torch.Tensor):
+            new_shape = tuple(size if is_symbolic(d) else int(d) for d in val.shape)
```
Comment (BoyuanFeng) [inline vllm/compilation/piecewise_backend.py]: should we assert there is only 1 symbolic shape in val.shape?

File: vllm/compilation/piecewise_backend.py [hunk 5]
```diff
@@ -16,14 +15,47 @@
...
+            args.append(torch.zeros(new_shape, dtype=val.dtype, device=val.device))
+        else:
+            args.append(val)
```
Comment (BoyuanFeng) [inline vllm/compilation/piecewise_backend.py:70]: should we be more explicit on the arg types? i.e., only allow SymInt and torch.Tensor.

File: vllm/compilation/backends.py [hunk 6]
```diff
@@ -1041,6 +1027,31 @@ def __call__(self, graph: fx.GraphModule, example_inputs: Sequence[Any]) -> Any:
             self.split_gm, submod_names_to_compile, self.vllm_config, self
         ).run(*fake_args)
 
+        # Compile all ranges for all piecewise backends up front so that
```
Comment (ProExpertProg) [inline vllm/compilation/backends.py]: Why can't we do this as part of the `PiecewiseCompileInterpreter.run`? That way we wouldn't have to manually find the piecewise backend instances on the graph.

**Review by ProExpertProg — APPROVED** [review]
Overall LGTM, I think @BoyuanFeng's comments are fair, I'll let you guys take this across the finish line

**Review by zhxchen17 — APPROVED** [review]

File: vllm/compilation/piecewise_backend.py (hunk 4 above)
Comment (zou3519) [reply vllm/compilation/piecewise_backend.py]: this is difficult to do. It's possible for val.shape to have 2 symints that are "different" (e.g. s0 vs s1) but later on we prove s0 == s1

File: vllm/compilation/piecewise_backend.py (hunk 5 above)
Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:70]: We support OpaqueObject and TorchBind objects so this is a bit tricky to allow.

Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:70]: I think I'll just handle the OpaqueObject case

Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:70]: Actually I am not going to handle this case because I don't know what OpaqueObject shows up as. Are you OK if I revisit when 2.11 releases? I don't want to make that upgrade more complicated by adding an incorrect assert right now.

**Review by BoyuanFeng — APPROVED** [review]

File: vllm/compilation/backends.py (hunk 6 above)
Comment (zou3519) [reply vllm/compilation/backends.py]: Hmm let me try this

Comment (zou3519) [reply vllm/compilation/backends.py]: I did this refactor, it worked out better, thanks for the suggestion

**Comment by zou3519** [comment]
@ProExpertProg @zhxchen17 I ended up needing to combine the log message into one: "torch.compile and initial profiling run took Xs in total". This is what the state was before this PR, so I think that's fine. I'll take a followup to actually split the messages.

Here's what's going on:
- in cold start we have a good split between compile vs the profiling run
- in warm start, we have a problem (with aot_compile). I had some trouble with getting the aot_compile to do all of the compilation during the compile step. Basically we need to load the dynamo artifact first, then look at the traced_files saved in it, and then do a compile run. Right now that compile run is inside the initial profiling run. It should be possible to continue refactoring this but this PR already changes a lot so to minimize risk I'd prefer to do more refactoring in a follow-up

Also, the "overlap weight load and compilation" work only really needed cold start to have a good split between compile vs profiling run, so shipping this PR by itself would unblock me from trying more things out there

File: vllm/compilation/piecewise_backend.py [hunk 7]
```diff
@@ -16,14 +15,62 @@
...
+        val = node.meta["example_value"]
+        if isinstance(val, torch.SymInt):
+            args.append(concretize(val))
+        elif isinstance(val, torch.Tensor):
+            new_shape = tuple(concretize(d) for d in val.shape)
+            new_strides = tuple(concretize(s) for s in val.stride())
+            # Use empty_strided to preserve the original tensor's layout
+            # (e.g. transposed weights), then zero out to avoid
+            # non-determinism from uninitialized memory.
+            t = torch.empty_strided(
```
Comment (eellison) [inline vllm/compilation/piecewise_backend.py]: do we need the storage_offset as well ? I imagine theres a helper somewhere to create a real tensor from fake in pytorch core. although it may be annoying to pin down across multiple pytorch versions.

File: vllm/compilation/piecewise_backend.py [hunk 8]
```diff
@@ -16,14 +15,62 @@
...
+                new_shape, new_strides, dtype=val.dtype, device=val.device
+            )
+            t.zero_()
```
Comment (eellison) [inline vllm/compilation/piecewise_backend.py]: just to double check : does `zero_` exist for low precision dtypes as well ? sometimes the fp8 and below dtypes are not well supported.

File: vllm/compilation/piecewise_backend.py [hunk 9]
```diff
@@ -216,27 +237,54 @@ def serialize(fn: Callable[..., Any]) -> bytes:
...
+        for range_entry in self.range_entries.values():
+            if range_entry.compiled:
+                continue
+
+            self._log_compile_start(range_entry.compile_range)
+
+            if range_entry.compile_range.is_single_size():
+                args_list = create_concrete_args(
+                    self.graph, range_entry.compile_range.start
+                )
```
Comment (eellison) [inline vllm/compilation/piecewise_backend.py:257]: Should we do this at the start of the range or the end ? I guess this is for eager compilation only. but even then it might be better to use the higher end.

File: vllm/compilation/piecewise_backend.py [hunk 10]
```diff
@@ -16,14 +15,62 @@
...
+    inputs but don't have real runtime tensors yet.
+    """
+    from torch.fx.experimental.symbolic_shapes import is_symbolic
+
+    def concretize(sym_val: Any) -> int:
+        """Replace all symbolic variables in a SymInt expression with size."""
+        if not is_symbolic(sym_val):
+            return int(sym_val)
+        expr = sym_val.node.expr
+        return int(expr.subs({s: size for s in expr.free_symbols}))
```
Comment (eellison) [inline vllm/compilation/piecewise_backend.py:50]: i suppose the single size is an invariant of vllm.

File: vllm/compilation/piecewise_backend.py (hunk 9 above)
Comment (ProExpertProg) [reply vllm/compilation/piecewise_backend.py:257]: Agree with end

File: vllm/compilation/piecewise_backend.py (hunk 8 above)
Comment (zou3519) [reply vllm/compilation/piecewise_backend.py]: welp, guess I'm using empty_strided

File: vllm/compilation/piecewise_backend.py (hunk 10 above)
Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:50]: at the moment yes

File: vllm/compilation/piecewise_backend.py (hunk 9 above)
Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:257]: for single_size start == end right?

Comment (zou3519) [reply vllm/compilation/piecewise_backend.py:257]: This is the case where the compile range is e.g. [8, 8]

File: vllm/compilation/piecewise_backend.py (hunk 8 above)
Comment (eellison) [reply vllm/compilation/piecewise_backend.py]: (i havent double checked this yet, just a thought)

Comment (zou3519) [reply vllm/compilation/piecewise_backend.py]: In theory the single-size compilation can just use FakeTensors. In practice that's another refactor that I don't want to do right now, but that will also supercede all of our concerns lol

Comment (eellison) [reply vllm/compilation/piecewise_backend.py]: okay, sorry, they do all support zero!
